workers/unity/Packages/io.improbable.worker.sdk/ParameterConversion.cs

```diff
@@ -0,0 +1,233 @@
+
+                    newEvent.UnixTimestampMillis = itemContainer->ItemUnion.Event.UnixTimestampMillis;
+                    newEvent.Type = ApiInterop.FromUtf8Cstr(itemContainer->ItemUnion.Event.Type);
+                    newEvent.Message = ApiInterop.FromUtf8Cstr(itemContainer->ItemUnion.Event.Message);
+
+                    // Copy the pointer to the underlying event data structure to get access to it's fields
+                    newEvent.Data = new TraceEventData(itemContainer->ItemUnion.Event.Data);
+                    var fields = newEvent.Data.GetAll();
+
+                    // Dispose of the copied internalContainer event data pointer as it's only guaranteed to exist
+                    // in this function scope
+                    newEvent.Data.EventData.Dispose();
+
+                    // Add the data to the newly initialized event data struct
+                    newEvent.Data = new TraceEventData();
+                    newEvent.Data.AddAll(fields);
+
+                    newItem.Event = newEvent;
+                    break;
+                default:
+                    throw new NotSupportedException("Invalid Item Type provided.");
+            }
+
+            return newItem;
+        }
+
+        public static unsafe void ConvertItem(Item item, ItemCallback callback)
+        {
+            var newItem = Item.GetThreadLocalItem();
+
+            newItem->ItemUnion = new CEventTrace.Item.Union();
+            newItem->ItemType = (CEventTrace.ItemType) item.ItemType;
+            switch (item.ItemType)
+            {
+                case ItemType.Span:
+                    var spanItem = item.Span.GetValueOrDefault();
+                    newItem->ItemUnion.Span = new CEventTrace.Span();
+                    newItem->ItemUnion.Span.Id = ConvertSpanId(spanItem.Id);
+                    newItem->ItemUnion.Span.CauseCount = (uint) spanItem.Causes.Length;
+                    for (var i = 0; i < newItem->ItemUnion.Span.CauseCount; i++)
+                    {
+                        newItem->ItemUnion.Span.Causes[i] = ConvertSpanId(spanItem.Causes[i]);
+                        callback(newItem);
+                    }
+
+                    break;
+                case ItemType.Event:
+                    var eventItem = item.Event.GetValueOrDefault();
+                    ConvertEvent(eventItem, internalEvent =>
+                    {
+                        newItem->ItemUnion.Event = internalEvent;
+                        callback(newItem);
+                    });
+
+                    break;
+                default:
+                    throw new NotSupportedException("Invalid Item Type provided.");
+            }
+        }
+
+        public static unsafe void ConvertEventTracerParameters(EventTracerParameters[] parameters, EventTracerParametersCallback callback)
+        {
+            var handles = new List<WrappedGcHandle>(parameters.Length);
```

Comment:
If this `List` is of a static size, might be better to use an array instead.

Reply:
Good point 👍